File: reference/typescript/src/schema/capability.ts
import { z } from 'zod/v4';
import { SPEC_VERSION } from './event.js';

// An audit capability object. It is read as a requirement when the host declares it and as a
// supported capability when the tool declares it. Where each travels is the binding's concern
// (§6.4, §6.5); what it holds and how two compare is §6.1's. spec_version carries the supported Auditable MCP version so a common
// version is established before events (which carry spec_version) are exchanged. All four fields
// are REQUIRED.
export const auditCapabilitySchema = z.strictObject({
  spec_version: z.string(), // supported Auditable MCP version, e.g. auditable-mcp/0.3
  level: z.enum(['L1', 'L2']),
  attempt: z.literal('request'), // attempt is always a blocking request (fail-closed)
  countersign: z.enum(['none', 'host']), // the host offers it, the tool requires it (§5.2)
});

export type AuditCapability = z.infer<typeof auditCapabilitySchema>;

export const DEFAULT_L1_CAPABILITY: AuditCapability = {
  spec_version: SPEC_VERSION,
  level: 'L1',
  attempt: 'request',
  countersign: 'none',
};

// The two axes run in opposite directions. On `level` the tool produces and the host requires, so
// an L2-capable tool satisfies an L1 requirement (a safe downgrade). On `countersign` (§5.2) the host
// produces and the tool requires, so a host offering `host` satisfies a tool requiring `none`.
const LEVEL_RANK: Record<string, number> = { L1: 1, L2: 2 };
const COUNTERSIGN_RANK: Record<string, number> = { none: 1, host: 2 };

// Why a call is or is not audit-negotiated (§6.2). `undeclared` is not a mismatch: nothing was
// offered to compare, and collapsing the two would brick the tool against ordinary MCP hosts.
export const NEGOTIATED = 'negotiated';
export const UNDECLARED = 'undeclared';
export const MISMATCH = 'mismatch';

export interface NegotiationResult {
  tool: AuditCapability;
  host: AuditCapability | undefined; // undefined when the peer declared no auditable-mcp extension
  outcome: typeof NEGOTIATED | typeof UNDECLARED | typeof MISMATCH;
  versionMatch: boolean;
  levelFit: boolean;
  countersignFit: boolean;
  negotiated: boolean; // true only for an audit-negotiated call; §6.2 governs every other case
}

// True if the tool offers at least the level the host requires (§6.1).
export function levelSatisfies(tool: AuditCapability, host: AuditCapability): boolean {
  return (LEVEL_RANK[tool.level] ?? 0) >= (LEVEL_RANK[host.level] ?? 0);
}

// True if the host provides at least the countersignature the tool requires (§5.2, §6.1).
export function countersignSatisfies(host: AuditCapability, tool: AuditCapability): boolean {
  return (COUNTERSIGN_RANK[host.countersign] ?? 0) >= (COUNTERSIGN_RANK[tool.countersign] ?? 0);
}

// Compare a host and a tool declaration (§6.1). A 0.x draft has no on-the-wire compatibility
// window, so a fit requires an exact spec_version match as well as both axes; the per-axis flags
// surface which one failed. Truthfulness is not verified here; runtime validation (§7) enforces the
// required level.
export function negotiateCapability(
  host: AuditCapability | undefined,
  tool: AuditCapability,
): NegotiationResult {
  if (host === undefined) {
    return {
      tool,
      host: undefined,
      outcome: UNDECLARED,
      versionMatch: false,
      levelFit: false,
      countersignFit: false,
      negotiated: false,
    };
  }
  const versionMatch = tool.spec_version === host.spec_version;
  const levelFit = levelSatisfies(tool, host);
  const countersignFit = countersignSatisfies(host, tool);
  const fits = versionMatch && levelFit && countersignFit;
  return {
    tool,
    host,
    outcome: fits ? NEGOTIATED : MISMATCH,
    versionMatch,
    levelFit,
    countersignFit,
    negotiated: fits,
  };
}

File: reference/python/src/auditable_mcp/capability.py
"""Audit capability object and the two axes negotiation compares (§6.1).

Read as a requirement when the host declares it and as a supported capability when the tool declares
it; both directions are exchanged during the MCP initialize phase.

The two axes run in opposite directions. On `level` the tool produces and the host requires, so a
tool offering L2 satisfies an L1 host. On `countersign` (§5.2) the host produces and the tool requires,
so a host offering `host` satisfies a tool that requires `none`. Each party enforces the axis on
which it is the one requiring.

A peer that declared nothing is not a failed negotiation but an absent one, which §6.2 governs
differently: the tool sends no audit message at all and serves the call as an ordinary MCP tool.
"""

from dataclasses import dataclass

SPEC_VERSION = 'auditable-mcp/0.3'


@dataclass(frozen=True)
class AuditCapability:
    """An audit capability: a requirement when host-declared, a supported level when tool-declared.

    spec_version carries the supported Auditable MCP version so a common version is established
    before events (which carry spec_version) are exchanged (§6.1). All four fields are REQUIRED.
    """

    spec_version: str = SPEC_VERSION  # supported version, e.g. auditable-mcp/0.3
    level: str = 'L1'  # 'L1' | 'L2'
    attempt: str = 'request'  # attempt is always a blocking request (fail-closed)
    countersign: str = 'none'  # 'none' | 'host' - the host offers it, the tool requires it (§5.2)


DEFAULT_L1_CAPABILITY = AuditCapability()

# L2 obligations are a superset of L1, so an L2-capable tool satisfies an L1 requirement (a safe
# downgrade), while an L1-only tool does not satisfy an L2 requirement.
_LEVEL_RANK = {'L1': 1, 'L2': 2}

# A host that signs satisfies a tool that requires a signature and one that does not; a host that
# does not sign satisfies only the latter.
_COUNTERSIGN_RANK = {'none': 1, 'host': 2}

# Why a session is or is not audit-negotiated (§6.2). UNDECLARED is not a mismatch: nothing was
# offered to compare, and collapsing the two would brick the tool against ordinary MCP hosts.
NEGOTIATED = 'negotiated'
UNDECLARED = 'undeclared'
MISMATCH = 'mismatch'


@dataclass(frozen=True)
class NegotiationResult:
    """The outcome of a capability exchange, and which axis decided it."""

    tool: AuditCapability
    host: AuditCapability | None
    outcome: str  # NEGOTIATED | UNDECLARED | MISMATCH
    version_match: bool
    level_fit: bool
    countersign_fit: bool

    @property
    def negotiated(self) -> bool:
        """True only for an audit-negotiated session; §6.2 governs every other case."""
        return self.outcome == NEGOTIATED


def level_satisfies(tool: AuditCapability, host: AuditCapability) -> bool:
    """Return True if the tool offers at least the level the host requires (§6.1)."""
    return _LEVEL_RANK.get(tool.level, 0) >= _LEVEL_RANK.get(host.level, 0)


def countersign_satisfies(host: AuditCapability, tool: AuditCapability) -> bool:
    """Return True if the host provides at least the countersign the tool requires (§5.2, §6.1)."""
    return _COUNTERSIGN_RANK.get(host.countersign, 0) >= _COUNTERSIGN_RANK.get(tool.countersign, 0)


def negotiate_capability(host: AuditCapability | None, tool: AuditCapability) -> NegotiationResult:
    """Compare a host and a tool declaration (§6.1).

    A 0.x draft has no on-the-wire compatibility window, so a fit requires an exact spec_version
    match as well as both axes; the per-axis flags surface which one failed. Truthfulness is not
    verified here; runtime validation (§7) enforces the required level.

    Args:
        host: The capability the host declared, or None if it declared no auditable-mcp extension.
        tool: The capability the tool declares.

    Returns:
        A result carrying both declarations, the outcome, and the per-axis fit.
    """
    if host is None:
        return NegotiationResult(
            tool=tool, host=None, outcome=UNDECLARED, version_match=False, level_fit=False, countersign_fit=False
        )
    version_match = tool.spec_version == host.spec_version
    level_fit = level_satisfies(tool, host)
    countersign_fit = countersign_satisfies(host, tool)
    fits = version_match and level_fit and countersign_fit
    return NegotiationResult(
        tool=tool,
        host=host,
        outcome=NEGOTIATED if fits else MISMATCH,
        version_match=version_match,
        level_fit=level_fit,
        countersign_fit=countersign_fit,
    )
